src/expr/transform/demand.rs

```diff
@@ -94,6 +101,23 @@ impl Demand {
                     columns.extend(new_columns.clone());
                     new_columns.retain(|c| *c >= arity);
                 }
+
+                // Replace un-read expressions with Null literals to prevent evaluation.
+                for (index, scalar) in scalars.iter_mut().enumerate() {
+                    if !columns.contains(&(arity + index)) {
+                        // Leave literals as they are, to benefit explain.
+                        if !scalar.is_literal() {
+                            let typ = relation_type.column_types[arity + index].clone();
+                            let datum = if typ.nullable {
+                                repr::Datum::Null
+                            } else {
+                                typ.scalar_type.dummy_datum()
```

Comment:
Can we not now just use null, seeing as we're removing all uses?

Or maybe a Datum::DontLookAtMe if we're worried about typing.

Reply:
The motivation here was to avoid antagonizing `typ()`, which might change its mind about whether the type was nullable or not based on now seeing a constant null there.

Reply:
Right, if we had a `Datum::DontLookAtMe` that would make sense. I didn't want to do something like that unilaterally. It may make sense in the future, or it may make more sense to perform whatever surgery is required to just remove the column.

Reply:
Or maybe just drop the ScalarExpr and fix up later references?

Reply:
That's a different approach to the same problem, I think. If we have both project lifting and literal map lifting, I think you can let that fix up the references (vs a bunch of surgery for each way they could be used).

Reply:
(Sorry about the non-sequiturs - your replies didn't show up for me until now)